src/value.rs

```diff
@@ -0,0 +1,553 @@
+            }
+            Value::Int64(i) => {
+                writer.write_safe_tag(b"i8", &i.to_string()[..])?;
+            }
+            Value::Bool(b) => {
+                writer.write_safe_tag(b"boolean", if b { "1" } else { "0" })?;
+            }
+            Value::String(ref s) => {
+                writer.write_tag(b"string", &s[..])?;
+            }
+            Value::Double(d) => {
+                writer.write_safe_tag(b"double", &d.to_string()[..])?;
+            }
+            Value::DateTime(date_time) => {
+                writer.write_safe_tag(b"dateTime.iso8601", &format!("{}", date_time)[..])?;
+            }
+            Value::Base64(ref data) => {
+                writer.write_safe_tag(b"base64", &encode_base64(data)[..])?;
+            }
+            Value::Struct(ref map) => {
+                writer.write_start_tag(b"struct")?;
+                for (ref name, ref value) in map {
+                    writer.write_start_tag(b"member")?;
+                    writer.write_tag(b"name", &name[..])?;
+                    writer
+                        .write(&value.stringify()?.as_ref())
+                        .map_err(EncodingError::from)?;
+                    writer.write_end_tag(b"member")?;
+                }
+                writer.write_end_tag(b"struct")?;
+            }
+            Value::Array(ref array) => {
+                writer.write_start_tag(b"array")?;
+                writer.write_start_tag(b"data")?;
+                for value in array {
+                    // Raw write the value to the buffer because it's encoded xml.
+                    writer
+                        .write(&value.stringify()?.as_ref())
+                        .map_err(EncodingError::from)?;
+                }
+                writer.write_end_tag(b"data")?;
+                writer.write_end_tag(b"array")?;
+            }
+            Value::Nil => {
+                writer.write(b"<nil />").map_err(EncodingError::from)?;
+            }
+        }
+        writer.write_end_tag(b"value")?;
+
+        Ok(String::from_utf8(buf).map_err(EncodingError::from)?)
+    }
+
+    /// Returns an inner struct or array value indexed by `index`.
+    ///
+    /// Returns `None` if the member doesn't exist or `self` is neither a struct nor an array.
+    ///
+    /// You can also use Rust's square-bracket indexing syntax to perform this operation if you want
+    /// a default value instead of an `Option`. Refer to the top-level [examples](#examples) for
+    /// details.
+    /*
+    pub fn get<I: Index>(&self, index: I) -> Option<&Value> {
+        index.get(self)
+    }
+    */
+
+    /// If the `Value` is a normal integer (`Value::Int`), returns associated value. Returns `None`
+    /// otherwise.
+    ///
+    /// In particular, `None` is also returned if `self` is a `Value::Int64`. Use [`as_i64`] to
+    /// handle this case.
+    ///
+    /// [`as_i64`]: #method.as_i64
+    pub fn as_i32(&self) -> Option<i32> {
```

Comment:
It'd be cool if this could be a generic function similar to str's parse() function.  Then you could take advantage of the turbofish syntax for inference.

But it looks like that parse() function explicitly relies on the existence of the FromStr trait.  So I think to do something similar, you would need a FromValue trait, and then implement that trait on all the types you want to support.  So might not actually be worth the extra effort (I'm not even sure if it's possible to extend primitives with custom traits anyway).

Reply:
I suppose these could be converted to TryFrom or TryInto

Reply:
Going to leave it for now